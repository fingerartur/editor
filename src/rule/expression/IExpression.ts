import { IArg } from './IArg';

export interface IExpression {
    what: 'expression';
    root: IArg;
}