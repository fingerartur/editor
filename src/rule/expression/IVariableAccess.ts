import { IValue } from './IValue';
import { IType } from '../types/IType';

export interface IVariableAccess extends IValue {
    id: number;
    access: string;
    type: IType;
}