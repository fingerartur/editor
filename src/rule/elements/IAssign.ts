import { IVariable } from '../expression/IVariable';
import { IExpression } from '../expression/IExpression';
import { IElement } from './IElement';

export interface IAssign extends IElement {
    variable: IVariable;
    expression: IExpression;
}