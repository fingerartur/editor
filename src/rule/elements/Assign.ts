import { AssignEnhancer } from 'enhancers/AssignEnhancer';
import { SubRule } from '../SubRule';
import { Cloner } from '../Cloner';
import { AssigningElement } from './AssigningElement';
import { IVariable } from '../expression/IVariable';
import { Expression } from '../expression/Expression';
import { IAssign } from './IAssign';

export class Assign extends AssigningElement {
    public variable: IVariable;
    public expression: Expression;
    
    constructor(
        parentRule: SubRule,
        serialization?: IAssign
    ) {
        super(parentRule);
        if (serialization) {
            this.deserialize(serialization);
        } else {
            this.initialize();
        }
    }

    getType(): string {
        return 'assign';
    }

    serialize(): IAssign {
        const result = super.serialize() as IAssign;
        result.variable = this.variable;
        result.expression = this.expression.serialize();
        return result;
    }

    enhance(): AssignEnhancer {
        return new AssignEnhancer(this);
    }

    clone(): Assign {
        return new Assign(null, Cloner.clone(this) as any);
    }

    getIdsOfVariablesOnTheLeft(): number[] {
        let result = [];
        if (this.variable && this.variable.id != null) {
            result.push(this.variable.id);
        }
        return result;
    }

    getIdsOfVariablesOnTheRight(): number[] {
        if (this.expression) {
            return this.expression.getIdsOfVariables();
        } else {
            return [];
        }
    }

    merge(assign: Assign): void {
        this.text = assign.text;
        this.variable = assign.variable;
        this.expression = assign.expression;
    }

    isDeclaration(): boolean {
        return this.variable && this.variable.isDeclaration;
    }

    replaceIds(map: Map<number, number>): void {
        if (this.variable) {
            const newId = map.get(this.variable.id);
            if (newId) {
                this.variable.id = newId;
            }
        }
        if (this.expression) {
            this.expression.replaceIds(map);
        }
    }

    deserialize(serialization: IAssign): void {
        super.deserialize(serialization);
        this.variable = serialization.variable;
        this.expression = new Expression(serialization.expression);
    }

    private initialize(): void {
        this.variable = null;
        this.expression = new Expression();
        this.text = 'Assign';
    }
}