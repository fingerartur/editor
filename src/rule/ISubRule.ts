import { IElement } from './elements/IElement';

export interface ISubRule {
    what: 'sub-rule';
    elements: IElement[];
}