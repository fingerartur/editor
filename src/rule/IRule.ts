import { ISubRule } from './ISubRule';
import { IDeclarations } from './declarations/IDeclarations';

export interface IRule {
    id: number;
    name: string;
    description: string;
    rule: ISubRule;
    declarations: IDeclarations;
}