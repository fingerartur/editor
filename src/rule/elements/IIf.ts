import { IElement } from './IElement';
import { IExpression } from '../expression/IExpression';
import { ISubRule } from '../ISubRule';

export interface IIf extends IElement {
    condition: IExpression;
    branchTrue: ISubRule;
    branchFalse: ISubRule;
}