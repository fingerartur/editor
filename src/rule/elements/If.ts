import { Element } from './Element';
import { SubRule } from '../SubRule';
import { Expression } from '../expression/Expression';
import { IIf } from './IIf';
import { Types } from '../types/Types';
import { IfEnhancer } from 'enhancers/IfEnhancer';
import { HardwiredTypes } from '../types/HardwiredTypes';
import { Cloner } from '../Cloner';

export class If extends Element {
    private condition: Expression;
    private branchTrue: SubRule;
    private branchFalse: SubRule;
    
    constructor(
        parentRule: SubRule,
        serialization?: IIf
    ) {
        super(parentRule);
        if (serialization) {
            this.deserialize(serialization);
        } else {
            this.initialize();
        }
    }

    setCondition(condition: Expression): void {
        const booleanType = { name: HardwiredTypes.BOOL };
        if (!Types.areEqual(condition.getType, booleanType) || !Types.areEqual(condition.getRoot().type, booleanType)) {
            throw Error('bug: condiiton must be of boolean type');
        }
        this.condition = condition;
    }

    getCondition(): Expression {
        return this.condition;
    }

    getBranchTrue(): SubRule {
        return this.branchTrue;
    }

    getBranchFalse(): SubRule {
        return this.branchFalse;
    }

    serialize(): IIf {
        const result = super.serialize() as IIf;
        result.condition = this.condition.serialize();
        result.branchTrue = this.branchTrue.serialize();
        result.branchFalse = this.branchFalse.serialize();
        return result;
    }

    getType(): string {
        return 'if';
    }

    getSubRules(): SubRule[] {
        return[ this.branchFalse, this.branchTrue ];
    }

    enhance(): IfEnhancer {
        return new IfEnhancer(this);
    }

    clone(): If {
        return new If(null, Cloner.clone(this) as any);
    }

    merge(ifElement: If): void {
        this.condition = ifElement.condition;
        this.text = ifElement.text;
    }

    getIdsOfVariablesOnTheLeft(): number[] {
        return [];
    }

    getIdsOfVariablesOnTheRight(): number[] {
        if (this.condition) {
            return this.condition.getIdsOfVariables();
        } else {
            return [];
        }
    }

    replaceIds(map: Map<number, number>): void {
        this.condition.replaceIds(map);
    }

    deserialize(serialization: IIf): void {
        super.deserialize(serialization);
        this.condition = new Expression(serialization.condition);
        this.branchTrue = new SubRule(this, serialization.branchTrue);
        this.branchFalse = new SubRule(this, serialization.branchFalse);
    }

    private initialize(): void {
        this.branchTrue = new SubRule(this);
        this.branchFalse = new SubRule(this);
        this.condition = new Expression({
            what: 'expression',
            root: {
                name: 'condition',
                type: { name: HardwiredTypes.BOOL, isMultiValued: false },
                value: null,
                error: null
            }
        });
        this.text = 'If';
    }
}