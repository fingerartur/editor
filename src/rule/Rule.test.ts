import { IVariableAccess } from './expression/IVariableAccess';
import { IExpression } from './expression/IExpression';
import { If } from './elements/If';
import { Assign } from './elements/Assign';
import { ISubRule } from './ISubRule';
import { IIf } from './elements/IIf';
import { IAssign } from './elements/IAssign';
import { IRule } from './IRule';
import { Rule } from './Rule';

const condition: IExpression = {
    what: 'expression',
    root: {
        name: 'root',
        type: { name: 'boolean' },
        value: {
            what: 'variableAccess',
            id: 1,
            access: 'globalVar',
            type: { name: 'boolean' }
        } as IVariableAccess
    }
};

const branchFalse: ISubRule = {
    what: 'sub-rule',
    elements: [
        { what: 'element', elementType: 'end', text: 'start' },
        { what: 'element', elementType: 'assign', text: 'assign to z', variable: { id: 1003, isDeclaration: true, name: 'z', type: { name: 'int'}}, expression: null } as IAssign,
        { what: 'element', elementType: 'end', text: 'end' }
    ]
};
const branchTrue: ISubRule = {
    what: 'sub-rule',
    elements: [
        { what: 'element', elementType: 'end', text: 'start' },
        { what: 'element', elementType: 'assign', text: 'assign to x', variable: { id: 1001, name: 'x', type: { name: 'int'}}, expression: null } as IAssign,
        { what: 'element', elementType: 'end', text: 'end' }
    ]
};
const serialization: IRule = {
    id: 1,
    name: 'Test rule',
    description: 'Used in a unit test',
    declarations: {
        nextId: 1004,
        declarations: [
            { id: 1, name: 'globalVar', type: { name: 'boolean' }},
            { id: 1001, name: 'x', type: { name: 'int' }},
            { id: 1002, name: 'y', type: { name: 'int' }},
            { id: 1003, name: 'z', type: { name: 'int' }},
        ]
    },
    rule: {
        what: 'sub-rule',
        elements: [
            { what: 'element', elementType: 'end', text: 'start' },
            { what: 'element', elementType: 'assign', text: 'assign to x', variable: { id: 1001, isDeclaration: true, name: 'x', type: { name: 'int'}}, expression: null } as IAssign,
            { what: 'element', elementType: 'assign', text: 'assign to y', variable: { id: 1002, isDeclaration: true, name: 'y', type: { name: 'int'}}, expression: null } as IAssign,
            { what: 'element', elementType: 'if', text: 'condition', condition, branchFalse, branchTrue } as IIf,
            { what: 'element', elementType: 'end', text: 'end' }
        ]
    }
};


describe('Rule', () => {

    let rule: Rule;
    beforeEach(() => {
        rule = new Rule(serialization);
    });

    it('constructs', () => {
        rule = new Rule();
        expect(rule.rule.isEmpty()).toBeTruthy();
        expect(rule.declarations.getAll()).toEqual([]);
        expect(rule.rule.getFirst().graphicalMetadata.isVisible).toBeTruthy();
        expect(rule.rule.getLast().graphicalMetadata.isVisible).toBeTruthy();
    });

    it('deserializes', () => {
        expect(rule.rule.isEmpty()).toBeFalsy();
        expect(rule.declarations.getAll().length).toBe(4);
        expect(rule.rule.getFirst().next instanceof Assign).toBeTruthy();
        expect(rule.rule.getFirst().next.next instanceof Assign).toBeTruthy();
        expect(rule.rule.getFirst().next.next.next instanceof If).toBeTruthy();
    });

    it('returns the first element', () => {
        expect(rule.rule.getFirst()).toBe(rule.getRoot());
    });

    it('returns the scope of an element', () => {
        const ifElement = rule.rule.getLast().prev;
        expect(ifElement instanceof If).toBeTruthy();
        const scope = rule.getScopeOf(ifElement);
        expect(Array.from(scope.forbiddenVariableNames)).toEqual([ 'z' ]);
        expect(scope.visibleDeclarations.map(declaration => declaration.name).sort()).toEqual([ 'globalVar', 'x', 'y' ]);
    });

    describe('returns declarations visible in the scope of a concrete element', () => {

        it('always returns globals', () => {
            let declarations = rule.getVisibleDeclarationsFor(rule.getRoot());
            expect(declarations.length).toBe(1);
            expect(declarations[0].name).toBe("globalVar");
        });

        it('does not return anything else', () => {
            const element = (rule.rule.getLast().prev as If).getBranchTrue().getLast();
            let names = rule.getVisibleDeclarationsFor(element).map(declaration => declaration.name).sort();
            expect(names).toEqual(['globalVar', 'x', 'y']);
        });
        
    });

    it('returns IDs of local declarations visible in the scope of a concrete element', () => {
        const element = (rule.rule.getLast().prev as If).getBranchTrue().getLast();
        let ids = rule.getVisibleLocalVarIds(element).sort();
        expect(ids).toEqual([ 1001, 1002 ]);
    });

    it('returns IDs of any variables used in any way in any element after or descending from the given element', () => {
        const ids = rule.getIdsOfVariablesBelow(rule.getRoot()).sort();
        expect(ids).toEqual([ 1, 1001, 1002, 1003 ]);
    });

    it('returns IDs of all declarations in elements after or descending from the given element', () => {
        const ids = rule.getIdsOfDeclarationsBelow(rule.getRoot()).sort();
        expect(ids).toEqual([ 1001, 1002, 1003 ]);
    });

    it('returns names of all declarations in elements after or descending from the given element', () => {
        const names = rule.getNamesOfDeclarationsBelow(rule.getRoot()).sort();
        expect(names).toEqual([ 'x', 'y', 'z' ]);
    });

    it('serializes', () => {
        expect(rule.serialize()).toMatchSnapshot();
        // cannot use .toMatchObject() here easily, because the result contains empty expressions instead of nulls
    });

    it('returns info', () => {
        expect(rule.getInfo()).toEqual({
            id: 1,
            name: 'Test rule',
            description: 'Used in a unit test'
        });        
    });
});