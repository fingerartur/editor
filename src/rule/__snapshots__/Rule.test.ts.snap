// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rule serializes 1`] = `
Object {
  "declarations": Object {
    "declarations": Array [
      Object {
        "id": 1,
        "name": "globalVar",
        "type": Object {
          "name": "boolean",
        },
      },
      Object {
        "id": 1001,
        "name": "x",
        "type": Object {
          "name": "int",
        },
      },
      Object {
        "id": 1002,
        "name": "y",
        "type": Object {
          "name": "int",
        },
      },
      Object {
        "id": 1003,
        "name": "z",
        "type": Object {
          "name": "int",
        },
      },
    ],
    "nextId": 1004,
  },
  "description": "Used in a unit test",
  "id": 1,
  "name": "Test rule",
  "rule": Object {
    "elements": Array [
      Object {
        "elementType": "end",
        "errors": Array [],
        "text": "start",
        "what": "element",
      },
      Object {
        "elementType": "assign",
        "errors": Array [],
        "expression": Object {
          "root": Object {
            "error": null,
            "name": "root",
            "type": null,
            "value": null,
          },
          "what": "expression",
        },
        "text": "assign to x",
        "variable": Object {
          "id": 1001,
          "isDeclaration": true,
          "name": "x",
          "type": Object {
            "name": "int",
          },
        },
        "what": "element",
      },
      Object {
        "elementType": "assign",
        "errors": Array [],
        "expression": Object {
          "root": Object {
            "error": null,
            "name": "root",
            "type": null,
            "value": null,
          },
          "what": "expression",
        },
        "text": "assign to y",
        "variable": Object {
          "id": 1002,
          "isDeclaration": true,
          "name": "y",
          "type": Object {
            "name": "int",
          },
        },
        "what": "element",
      },
      Object {
        "branchFalse": Object {
          "elements": Array [
            Object {
              "elementType": "end",
              "errors": Array [],
              "text": "start",
              "what": "element",
            },
            Object {
              "elementType": "assign",
              "errors": Array [],
              "expression": Object {
                "root": Object {
                  "error": null,
                  "name": "root",
                  "type": null,
                  "value": null,
                },
                "what": "expression",
              },
              "text": "assign to z",
              "variable": Object {
                "id": 1003,
                "isDeclaration": true,
                "name": "z",
                "type": Object {
                  "name": "int",
                },
              },
              "what": "element",
            },
            Object {
              "elementType": "end",
              "errors": Array [],
              "text": "end",
              "what": "element",
            },
          ],
          "what": "sub-rule",
        },
        "branchTrue": Object {
          "elements": Array [
            Object {
              "elementType": "end",
              "errors": Array [],
              "text": "start",
              "what": "element",
            },
            Object {
              "elementType": "assign",
              "errors": Array [],
              "expression": Object {
                "root": Object {
                  "error": null,
                  "name": "root",
                  "type": null,
                  "value": null,
                },
                "what": "expression",
              },
              "text": "assign to x",
              "variable": Object {
                "id": 1001,
                "name": "x",
                "type": Object {
                  "name": "int",
                },
              },
              "what": "element",
            },
            Object {
              "elementType": "end",
              "errors": Array [],
              "text": "end",
              "what": "element",
            },
          ],
          "what": "sub-rule",
        },
        "condition": Object {
          "root": Object {
            "error": null,
            "name": "root",
            "type": Object {
              "name": "boolean",
            },
            "value": Object {
              "access": "globalVar",
              "id": 1,
              "type": Object {
                "name": "boolean",
              },
              "what": "variableAccess",
            },
          },
          "what": "expression",
        },
        "elementType": "if",
        "errors": Array [],
        "text": "condition",
        "what": "element",
      },
      Object {
        "elementType": "end",
        "errors": Array [],
        "text": "end",
        "what": "element",
      },
    ],
    "what": "sub-rule",
  },
}
`;
